sys/mips/include/pcpu.h

```diff
@@ -36,31 +36,57 @@
 #include <machine/cpufunc.h>
 #include <machine/pte.h>
 
+#include <cheri/cheric.h>
+
 #if defined(MIPS_EXC_CNTRS)
 #define	PCPU_MIPS_COUNTERS						\
 	register_t	pc_tlb_miss_cnt;	/* TLB miss count */    \
 	register_t	pc_tlb_invalid_cnt;	/* TLB invalid count */ \
 	register_t	pc_tlb_mod_cnt;		/* TLB modification count */ \
 #define	PCPU_NUM_EXC_CNTRS	3
 #else
-#define PCPU_MIPS_COUNTERS
+#define	PCPU_MIPS_COUNTERS
 #define	PCPU_NUM_EXC_CNTRS	0
 #endif
 
+#if defined(__CHERI_PURE_CAPABILITY__)
+/*
+ * Add an extra MD PCPU field for a cached copy of the current thread
+ * kernel stack capability.
+ * This must be kept in sync when context switching.
+ */
+#define	PCPU_MD_CAPABILITY_FIELDS					\
+	void	*pc_kstack_cap;		/* cached curthread kstack capability */
```

Comment:
Looking at the non-purecap exception entry points I fail to understand how on earth they manage to get the kernel stack without something like this (and RISC-V instead does things the other way round where it just spills the PCPU pointer to the stack on the way out to userspace and saves the kernel stack in a system register, so on the way into the kernel from userspace it has the stack immediately to hand for saving the trapframe then later loads up the PCPU).